### PR TITLE
Add options to avoid atexit, and to avoid pinning the libraries in memory

### LINKED ISSUE #7598: dll can't be free after it call bio_new and bio_free APIs in windows platform
Hi Team
I meet a strange problem when i use Openssl library.

Background
My project need to upgrade a DLL dynamically, so in application, it will load the DLL first and free it after application get a new version DLL. Then application will load the new DLL and run by new DLL.
But now, **my application can't free the DLL after it call bio_new and bio_free APIs. It's still in loading modules.**
If i don't call bio_new and bio_free，my application can free the DLL success.

Openssl is complied to a static library and linked in this DLL.
**I want to know if bio_new will use some system resource (file handle/network/memory) which will not be freed by bio_free.**
I doubt those resources block application to free DLL.

My Openssl version is openssl-1.1.0h.
Thanks
Forston

### PR DESCRIPTION
**Pull request by mattcaswell**
This PR introduces a compile time option and a run time option to control the clean up behaviour. At the moment we attempt to "pin" OpenSSL in memory until process exit. This is because of our use of the "atexit" handler. This handler is defined to run at process exit - but it can cause crashes if the library has already been unloaded at process exit. Some platforms (such as Linux) run the atexit handler when the library gets unloaded instead of at process exit.

We add two options to control this behaviour. Firstly a compile time option "no-pinso" which stops the so from being pinned in memory until process exit. Secondly a run time option OPENSSL_INIT_NO_ATEXIT which prevents the atexit handler from being registered. Using this option will require the application to clean up OpenSSL explicitly by using OPENSSL_cleanup().

As part of this I made 2 other related changes: Firstly fixed a bug in the `DSO_REFTEST` test in shlibloadtest, which effectively meant that the test was never run. Secondly I rewrote shlibloadtest to NOT use the test framework. The reason for this is that the test framework implicitly requires libcrypto to be linked (it uses BIOs). However the point about shlibloadtest is to test dynamic loading and unloading of the library. I am concerned that linking directly against libcrypto and also dynamically loading it might mask certain failure modes.

This PR supersedes and replaces #7388.

Fixes #7598 

It's not clear to me which branches this should go to. Technically these are features. You could also argue that the inability to control this is a bug (see #7598).

### CODE REVIEW

**Comment by mattcaswell** [comment]
@levitte - there might be some tweaking required to get this to work properly on VMS

**Comment by levitte** [comment]
> @levitte - there might be some tweaking required to get this to work properly on VMS

I doubt it, as unloading isn't supported as far as I know. I will double check that this hasn't changed in later years...

**Comment by mattcaswell** [comment]
The only reason I said so was because I wasn't sure about the purpose of these lines:

https://github.com/openssl/openssl/blob/6b956fe77b8aeb899ef7bdfa147a00bda51b804a/Configurations/10-main.conf#L1728-L1733

**Comment by mattcaswell** [comment]
Possibly its just to avoid the leak-a-reference trick because it isn't needed on VMS - in which case its fine and can stay.

**Comment by levitte** [comment]
> Possibly its just to avoid the leak-a-reference trick because it isn't needed on VMS - in which case its fine and can stay.

Yup, that is exactly the purpose

**Comment by richsalz** [comment]
I think using 'no-pinso' as the config name and NO_ATEXIT as the define are confusing.

File: INSTALL [hunk 1]
```diff
@@ -402,6 +402,23 @@
   no-pic
                    Don't build with support for Position Independent Code.
 
+  no-pinso         By default OpenSSL will attempt to stay in memory until the
```
Comment (mspncp) [inline INSTALL]: The abbreviation `no-pinso` is rather mysterious and not everybody will know right away that it stands for "don't pin the shared library".

**Comment by mattcaswell** [comment]
> I think using 'no-pinso' as the config name and NO_ATEXIT as the define are confusing.

Why? They do completely different things. They are not equivalent.

**Comment by mspncp** [comment]
> I think using 'no-pinso' as the config name and NO_ATEXIT as the define are confusing.
@richsalz the option suppresses the pinning of the shared library, not the atexit handler (see my first brief review comment).

**Comment by mattcaswell** [comment]
I could go with `no-pinshared`.

DLL pinning is implemented on Windows.

**Comment by mspncp** [comment]
> DLL pinning is implemented on Windows.

Ok. And what do you think about my suggestion to add a `DLL_PROCESS_DETACH` handler?

**Comment by mattcaswell** [comment]
> Ok. And what do you think about my suggestion to add a DLL_PROCESS_DETACH handler?

Windows has an `_onexit()` (as opposed to `atexit()`) which gets called on unload. We might want to consider using that instead.

**Comment by richsalz** [comment]
I understand they are different, but not completely separable.  At any rate, "pinshared" is a big improvement, thankas.

**Comment by mspncp** [comment]
> Windows has an _onexit() (as opposed to atexit()) which gets called on unload. We might want to consider using that instead.

Ah, thanks! I wasn't aware of `_onexit()`, that sounds much better and easier to use.

**Comment by levitte** [comment]
Why using `_onexit` is easier than adding another case in `DllMain`, I don't understand.

Do we know how long `_onexit` has existed?

**Comment by mspncp** [comment]
Because you would need both an `atexit()` handler and `DLL_PROCESS_DETACH` handler, since the latter is called only during dll unload, while the former at process exit. And if the dll gets unloaded will be in trouble to deregister the atexit handler. Or am I mistaken?

**Comment by levitte** [comment]
I'm not sure why we would have *both* `atexit` and whatever-shared-library-unload-mechanism at the same time.  We used `atexit` because we feared there was no consistently supportable on-shared-unload mechanism available for us.  Using `atexit` from inside a library is a pretty poor solution, really.

**Comment by levitte** [comment]
The main reason is that we use `atexit` to register automatic cleanup with `OPENSSL_cleanup`. Since `atexit` handlers traditionally called at process rundown, we need to ensure `OPENSSL_cleanup` sticks around that long.

**Comment by mattcaswell** [comment]
> Why using _onexit is easier than adding another case in DllMain, I don't understand.

Because its interface is practically the same as atexit which we are already using and keeps all the logic together in a single file which should lead to a more maintainable solution.

> Do we know how long _onexit has existed?

AFAICT it has been around since at least Windows XP (which is as far back as we care about).

**Comment by mattcaswell** [comment]
Please ignore the latest push...it includes test changes to try and figure out why MacOS is unhappy.

**Comment by mattcaswell** [comment]
I finally figured out why this was failing on MacOS. It turns out I was hitting a latent bug in our RUN_ONCE code. This bug probably also impacts 1.1.1 and maybe 1.1.0 too. Anyway travis is now passing.

This incorporates the same bug fix included in #7662. What I'm not sure about is why this passes but #7662 fails.

**Comment by mattcaswell** [comment]
Ping?

**Comment by mattcaswell** [comment]
I rebased this to fix a conflict with master. Still looking for a reviewer for this.

File: include/internal/thread_once.h [hunk 2]
```diff
@@ -30,6 +30,14 @@
     }                                           \
     static int init(void)
 
+#define DEFINE_RUN_ONCE_STATIC_ALT(initalt, init) \
+    static int initalt(void);                     \
+    static void initalt##_ossl_(void)             \
+    {                                             \
+        init##_ossl_ret_ = initalt();             \
```
Comment (levitte) [inline include/internal/thread_once.h]: Shouldn't there be a `static int init##_ossl_ret_ = 0;` somewhere in this macro definition?

File: include/internal/thread_once.h [hunk 3]
```diff
@@ -43,3 +51,6 @@
  */
 #define RUN_ONCE(once, init)                                            \
     (CRYPTO_THREAD_run_once(once, init##_ossl_) ? init##_ossl_ret_ : 0)
+
+#define RUN_ONCE_ALT(once, initalt, init)                               \
```
Comment (levitte) [inline include/internal/thread_once.h]: Please document, the same way `RUN_ONCE` is documented.  Ideally, all the `DEFINE_RUN_ONCE` and `DECLARE_RUN_ONCE` variants should be similarly documented, but that might be for another PR.

**Comment by levitte** [comment]
> Still looking for a reviewer for this.

Sorry that it took a while, it took me a bit to wrap my head around this PR

File: INSTALL [hunk 4]
```diff
@@ -411,6 +411,23 @@
...
+                   rather than at process exit. This option can be used to stop
+                   OpenSSL from attempting to stay in memory until the process
+                   exits. This could lead to crashes if either libcrypto or
+                   libssl have already been unloaded at the point
+                   that the atexit handler is invoked, e.g. on a platform which
+                   calls atexit() on unload of the library, and libssl is
+                   unloaded before libcrypto then a crash is likely to happen.
+                   Applications can suppress running of the atexit() handler at
+                   run time by using the OPENSSL_INIT_NO_ATEXIT option to
+                   OPENSSL_init_crypto(). See the man page for further details.
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply INSTALL]: Done

File: include/internal/thread_once.h (hunk 3 above)
Comment (mattcaswell) [reply include/internal/thread_once.h]: Done - I documented all of them.

File: include/internal/thread_once.h (hunk 2 above)
Comment (mattcaswell) [reply include/internal/thread_once.h]: No. That's kind of the point of introducing this variant. This fixes a bug in the RUN_ONCE code. Previously we had the concept of declaring more than one variant of an initialiser function where all the variants all shared the same CRYPTO_ONCE object - so only one variant would ultimately get run (dependant on the options you choose at run time). Unfortunately each variant had their own `init##_ossl_ret` variable. So if you called `RUN_ONCE(myaltinit)` followed later by `RUN_ONCE(myinit)`, the later `RUN_ONCE` call always returned 0 - because the return value for that alternative had never been set. This is incorrect behaviour. I've tried to explain it in the comments. Please take a look.

**Comment by mattcaswell** [comment]
Fixup commits pushed.

File: include/internal/thread_once.h (hunk 2 above)
Comment (vdukhovni) [reply include/internal/thread_once.h]: Agreed. The RUN_ONCE changes are correct.  Have not looked at the rest of this PR.

File: include/internal/thread_once.h [hunk 5]
```diff
@@ -30,6 +64,46 @@
...
 
+/*
+ * DEFINE_RUN_ONCE_STATIC_ALT: Define an alternative initialiser function. This
+ * function will be declared as static within the file. It takes no arguments
+ * and returns and int result (1 for success or 0 for failure). An alternative
+ * initialiser function is expected to be associated with a primary initialiser
+ * function defined via DEFINE_ONCE_STATIC where both functions use the same
+ * CRYPTO_ONCE object to synchronise. Where an alternative initialiser function
+ * is used only one of the primary or the alternative initialiser function will
+ * ever be called - and that function will be called excatly once. Definitition
```
Comment (vdukhovni) [inline include/internal/thread_once.h]: s/excatly/exactly/

File: include/internal/thread_once.h [hunk 6]
```diff
@@ -30,6 +64,46 @@
...
+ * DEFINE_RUN_ONCE_STATIC(myinitfunc)
+ * {
+ *     do_some_initialisation();
+ *     if (init_is_successful())
+ *         return 1;
+ *
+ *     return 0;
+ * }
+ * 
+ * DEFINE_RUN_ONCE_STATIC_ALT(myaltinitfunc)
```
Comment (vdukhovni) [inline include/internal/thread_once.h]: Takes a second parameter not mentioned in the comment.

**Comment by vdukhovni** [comment]
This PR looks fine to me.  No problems.
However, while this PR overlaps with #7969, each also addresses separate concerns in the same set of initialization functions.  At least in master the two should be combined.

**Comment by vdukhovni** [comment]
> @t-j-h This should go into master and 1.1.1 (it is fixing a defect in the initialisation approach)
In that case, I'd really like see this followed up by #7969 in both branches. I can rebase #7969 once this is merged, or even before, if there's any advantage to a consolidated PR.

**Comment by t-j-h** [comment]
I would suggest rebase after merge - but I'd leave that up to you and @mattcaswell to figure out.
I do see these as solving different problems where #7969 is about the higher-level conf issues and this one is purely about the atexit challenges. They don't belong together in a single PR IMHO.

**Comment by vdukhovni** [comment]
> @t-j-h I would suggest rebase after merge 
No worries, it is only by accident that they don't overlap more, because my original version also had the fixes for the RUN_ONCE barrier pairs, but then I wasn't sure whether failure on flag reversal was deliberate, so left it out for later... I'd also prefer to see downstream maintainers see these as a single set of fixes that address various limitations in the initialization code.

**Comment by mspncp** [comment]
(@vdukhovni in case you wonder why I edited your posts: I only fixed the markdown by inserting a blank line between the quote and your reply)

**Comment by mattcaswell** [comment]
I pushed this to master (including the trivial fixes suggested by @vdukhovni).

I have NOT pushed this to 1.1.1 yet due to non-trivial merge conflicts. I will open a separate PR for that branch.
